Issue 3770 transform phmsagas data

seeess1 opened the pull request:
# Overview

This is still WIP!

Relates to #3770.

## What problem does this address?

Goal of the PR is to complete the first transformation of raw PHMSA data into a core asset.

## What did you change?

- Added a new transform script for PHMSA data.
- Made one helper method (`fix_eia_na`) more generic since it can be applied across data sets (and updated references accordingly).
- Added a few other helper methods.
- Updated column map files with 2023 values (pulled from whatever was in 2022).
- Specifically changed the ordering of columns in `/Users/sam/Documents/pudl/src/pudl/package_data/phmsagas/column_maps/yearly_distribution.csv` since we had fax columns mapped to emails and vice versa.

Will come back to everything below once this draft PR has gone through an initial round of review.

# Documentation

Make sure to update relevant aspects of the documentation.

```[tasklist]
- [ ] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/nightly/release_notes.html): reference the PR and related issues.
- [ ] Update relevant Data Source jinja templates (see `docs/data_sources/templates`).
- [ ] Update relevant table or source description metadata (see `src/metadata`).
- [ ] Review and update any other aspects of the documentation that might be affected by this PR.
```

# Testing

## How did you make sure this worked? How can a reviewer verify this?

```[tasklist]
# To-do list
- [ ] If updating analyses or data processing functions: make sure to update or write data validation tests (e.g. `test_minmax_rows()`).
- [ ] Run `make pytest-coverage` locally to ensure that the merge queue will accept your PR.
- [ ] Review the PR yourself and call out any questions or issues you have.
- [ ] For minor ETL changes or data additions, once `make pytest-coverage` passes, make sure you have a fresh full PUDL DB downloaded locally, materialize new/changed assets and all their downstream assets and [run relevant data validation tests](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/testing.html#data-validation) using `pytest` and `--live-dbs`.
- [ ] For bigger ETL or data changes run the full ETL locally and then [run the data validations](https://catalystcoop-pudl.readthedocs.io/en/nightly/dev/testing.html#data-validation) using `make pytest-validate`.
- [ ] Alternatively, run the `build-deploy-pudl` GitHub Action manually.
```

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,164 @@
+    # (on the scale of thousands) do not actually sum up to "excavation_damage_total".
+    df[YEARLY_DISTRIBUTION_OPERATORS_COLUMNS["columns_with_nas_to_fill"]] = df[
+        YEARLY_DISTRIBUTION_OPERATORS_COLUMNS["columns_with_nas_to_fill"]
+    ].fillna(0)
```
We distinguish between NA as "not reported" and 0 as "reported as 0", so filling this would actually change the meaning here (from "I didn't fill this in" to "I had 0 incidents"). Please drop this step.

seeess1 replied:
Updated

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,164 @@
+    )
+
+    # Fill NA values with zeroes because these columns are simply counts.
+    # Note that "excavation_damage..." columns should sum up to the value in "excavation_damage_total". However, many rows
```
We should definitely add this validation as an asset check, and use fillna(0) there to do the comparison calculation.

seeess1 replied:
Got it. I'm seeing that even if I fill in NAs with zero, there are almost 4k rows where the `excavation_damage...` values don't sum to `excavation_damage_total`. How would you recommend handling these?

e-belfer reviewed (commented):
This is looking really good! A few comments on what's already been done, you're making great progress.

Some next steps:
- the values in `percent_unaccounted_for_gas` look like they could use some normalization (I see some negative values)
- once you're happy with the table, we can go ahead and write it to the database. See [here](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#update-the-pudl-db-schema) for guidance on how to do this.

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -1009,19 +1009,19 @@ def remove_leading_zeros_from_numeric_strings(
 
-def fix_eia_na(df: pd.DataFrame) -> pd.DataFrame:
+def fix_na(df: pd.DataFrame) -> pd.DataFrame:
     """Replace common ill-posed EIA NA spreadsheet values with np.nan.
```
If we're changing this function name, maybe something more descriptive like `standardize_na_values()` would be better than "fix" since I could see fixing NA values meaning filling them in with non-null values.

seeess1 replied:
Updated

zaneselvans commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,164 @@
+    ].apply(lambda col: col.str.title())
+
+    # List of state columns to standardize
+    df = standardize_state_columns(
```
I think we've used an `enum` constraint for the state column in EPA CEMS because it's a huge table and we want it to be interpreted as a categorical value to save space, and we don't use foreign key constraints in the hourly data that's only written to parquet. But for tables going into the database I think we would constrain the state values with a foreign key constraint that points at the `core_pudl__codes_subdivisions` table, and use the encoder setup to do the cleaning.

seeess1 replied:
Getting rid of my duplicate function and utilizing `POLITICAL_SUBDIVISIONS`. Plus adding an `enum` constraint to "state" now.

seeess1 replied:
Updated

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2185,305 @@ def retry(
+    }
+
+    for col in state_columns:
+        df[col] = df[col].replace(state_abbreviations).str.upper()
```
When you have a complete mapping between old and new values, the [col.map()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) method should be much faster. However any value that appears in the series but not in the dictionary defining the mapping gets converted to NA.

seeess1 replied:
Updated logic for standardizing state abbreviations here:

https://github.com/catalyst-cooperative/pudl/pull/3929/files#diff-941a3989c384fa527b5188c725a5b3beb1f1e0dbf9d427dae9dd82f4f9fe5c09R126

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2185,305 @@ def retry(
+
+
+def standardize_phone_column(df: pd.DataFrame, columns: list[str]) -> pd.DataFrame:
+    """Standardize phone numbers in the specified columns of the DataFrame.
+    US numbers: ###-###-####
+    International numbers with the international code at the beginning.
+    Numbers with extensions will be appended with "x#".
+    Non-numeric entries will be returned as np.nan. Entries with fewer than
+    10 digits will be returned with no hyphens.
+
+    Args:
+    df: The DataFrame to modify.
+    columns: A list of the names of the columns that need to be standardized
+    Returns:
+    The modified DataFrame with standardized phone numbers in the same column.
+    """
```
This docstring won't render correctly in the docs. Check how it looks in the built API docutmentation.

```suggestion
    """Standardize phone numbers in the specified columns of the DataFrame.
    
    US numbers: ###-###-####
    International numbers with the international code at the beginning.
    Numbers with extensions will be appended with "x#".
    Non-numeric entries will be returned as np.nan. Entries with fewer than
    10 digits will be returned with no hyphens.

    Args:
        df: The DataFrame to modify.
        columns: A list of the names of the columns that need to be standardized

    Returns:
        The modified DataFrame with standardized phone numbers in the same column.
    """
```

seeess1 replied:
Updated

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2185,305 @@ def retry(
+
+
+def analyze_missing_values(df: pd.DataFrame) -> list[str]:
+    """Analyze columns of a DataFrame for missing or invalid values. Note that this is purely for analysis
+    and does not perform any data transformation or cleaning.
+    This function checks each column for missing or custom missing values and prints
+    a summary of the findings for string (object), numeric, and datetime columns.
```
```suggestion
    """Analyze columns of a DataFrame for missing or invalid values.
    
    Note that this is purely for analysis and does not perform any data
    transformation or cleaning. This function checks each column for
    missing or custom missing values and prints a summary of the findings
    for string (object), numeric, and datetime columns.
```

seeess1 replied:
Updated

seeess1 commented:
> This is looking really good! A few comments on what's already been done, you're making great progress.
> 
> Some next steps:
> 
> * the values in `percent_unaccounted_for_gas` look like they could use some normalization (I see some negative values)
> * once you're happy with the table, we can go ahead and write it to the database. See [here](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#update-the-pudl-db-schema) for guidance on how to do this.

Ah yep I meant to point that part out about `percent_unaccounted_for_gas`. I noticed that too. I honestly wasn't sure how we wanted to handle this one. Any suggestions?

seeess1 commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2185,305 @@ def retry(
+
+        # Remove unwanted characters (parentheses, spaces, periods, and dashes) from the main phone number
+        phone_main = re.sub(
+            r"[^\d]", "", phone_main.replace(".0", "")
```
There were lots of phone numbers in the raw data with ".0" at the end of the value. So I'm using `replace` to remove those first.

zaneselvans replied:
You might want to use a regex that only matches this pattern at the end of the string to avoid accidentally removing characters from the middle of a number that could be reported like 303.443.0123

seeess1 replied:
Good point. I'm removing that logic. The regex should be all that's needed here.

seeess1 replied:
Actually I'm gonna leave in one line toward the top to get rid of these substrings just to make sure the rest of the number standardization is working with cleaner data.

Updated.

seeess1 commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,164 @@
+@asset(
+    ins={"raw_data": AssetIn(key="raw_phmsagas__yearly_distribution")},
+    io_manager_key=None,  # TODO: check on this one ("pudl_io_manager"? something else?)
+    compute_kind=None,  # TODO: check on this one
```
Good to know. I'll play around with this a little bit more.

seeess1 commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,164 @@
+    # Ensure all "report_year" values have four digits
+    mask = df["report_year"] < 100
+
+    # Convert 2-digit years to appropriate 4-digit format (assume cutoff at year 50)
```
No values below 50 in the data right now. I mostly put this in to keep it flexible enough to also be able to handle post-2000 data if we started seeing values like "00", "01", etc. The column still feels pretty reliable to me overall. It was just a 2 vs 4 digit year formatting issue.

zaneselvans commented:
Would `percent_unaccounted_for_gas < 0` mean that extra gas appeared from somewhere?

Is the distribution of the absolute values of positive and negative values similar?

Does it seem likely that different respondents just assumed different sign conventions here? Can we see what question they're responding to and if it is phrased ambiguously? Like is "5% of the gas was unaccounted for" supposed to be -5% or 5%?

seeess1 commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2185,305 @@ def retry(
+    return exception_cols
+
+
+def standardize_state_columns(df: pd.DataFrame, state_columns: list) -> pd.DataFrame:
```
Updated

zaneselvans commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,199 @@
+    )
+
+    # Function for looking up state names and abbreviations
+    def standardize_state(state):
+        if pd.isna(state):
+            return state
+        state = state.strip()
+        standardized_state = state_to_abbr.get(state, state)
+        if standardized_state not in state_to_abbr.values():
+            return np.nan
+        return standardized_state
```
Also note that you need to be careful to avoid mixing incompatible NA types.  `np.nan` is only appropriate for numpy floating point values. If it gets mixed in with other types in a column, the whole column has to be treated as an `object`.

For all the nullable Pandas dtypes (e.g. `Int64` and `string`) the appropriate NA value will be `pd.NA`.

seeess1 replied:
Ah whoops thanks for catching that. Fixing that now.

seeess1 replied:
Updated

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2194,219 @@ def retry(
+    """
+
+    # Function to clean and standardize phone numbers
+    def standardize_phone_number(phone: str) -> str:
```
See my related comment in the PHMSA transform module.

seeess1 replied:
Thanks for pointing this out. Logic has been updated.

zaneselvans commented on src/pudl/helpers.py:
```diff
@@ -2185,3 +2194,219 @@ def retry(
+    nan_cols = []
+    exception_cols = []
+
+    # Define custom missing value markers
+    custom_missing_values = [
+        "",
+        " ",
+        "NA",
+        "N/A",
+        "NULL",
+        "-",
+        "None",
+        "NaN",
+        "?",
+        "*",
+        "#",
+    ]
```
If this function is going to stick around post-PR then we probably want the user to be able to define their own set of custom missing values -- though these could be retained as a default set.

zaneselvans replied:
Probably also want to switch all of the `print()` statements to `logger.info()` outputs so it can be managed along with all of the other outputs.

seeess1 replied:
Updated

seeess1 commented on src/pudl/transform/phmsagas.py:
```diff
@@ -165,10 +172,10 @@ def standardize_state(state):
+    # Strip whitespace from all object (string) columns in the DataFrame
+    df[df.select_dtypes(include=["object"]).columns] = df.select_dtypes(include=["object"]).apply(
+        lambda col: col.map(lambda x: x.strip() if isinstance(x, str) else x)
+    )
```
It's been running very quickly on my dinky computer (less than half a second).

seeess1 commented:
> Would `percent_unaccounted_for_gas < 0` mean that extra gas appeared from somewhere?
> 
> Is the distribution of the absolute values of positive and negative values similar?
> 
> Does it seem likely that different respondents just assumed different sign conventions here? Can we see what question they're responding to and if it is phrased ambiguously? Like is "5% of the gas was unaccounted for" supposed to be -5% or 5%?

Here's the part of the form where this is coming from (from the [2004-2009 zip file](https://zenodo.org/records/13149085)):

<img width="433" alt="image" src="https://github.com/user-attachments/assets/7c52fa3e-aa20-44c1-861b-f9a35d2d9cdb">

Maybe negative values are coming from "appropriate adjustments"? That part's pretty vague. 

In either case, only 3.37% of values in this column are negative. I'm seeing negative values from across many different years, operators, states, etc. I was wondering if there might be a clear pattern here (one operator continually filing negative values for some reason) but so far I'm not seeing that. Can keep digging if needed.

seeess1 commented:
@e-belfer @zaneselvans I've responded to the initial round of comments. I still haven't solved the riddle of the negative unaccounted gas values. And I need to take a closer look at `io_manager_key` and `compute_kind`. But hopefully the PR is starting to shape up here.

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,159 @@
+def core_phmsagas__yearly_distribution_operators(
+    raw_data: pd.DataFrame,
+) -> pd.DataFrame:
+    """Build the :ref:`core_phmsagas__yearly_distribution_operators`."""
```
Here, include a short bulleted list of the transforms you make in this function so that this shows up in our docs. See pudl.transform.eia860 for some examples.

seeess1 replied:
Updated

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,159 @@
+    ]
+
+    # Fill NaN values with pd.NA, then cast to "Int64" nullable integer type
+    df[cols_to_convert] = df[cols_to_convert].fillna(pd.NA).astype("Int64")
```
Non-blocking: `convert_dtypes()` might more simply achieve what you're trying to do here, since it handles the conversion of nulls to Pandas nullable dtypes.

seeess1 replied:
Updated

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,159 @@
+        "preparer_phone",
+        "preparer_title",
+    ],
+    "columns_to_convert_to_ints": [
```
`federal_land_leaks_repaired_or_scheduled` should also be in this list, I believe!

seeess1 replied:
Good catch

seeess1 replied:
Updated

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -0,0 +1,159 @@
+        "services_shutoff_valve_in_system",
+        "services_shutoff_valve_installed",
+    ],
+    "capitalization_exclusion": ["headquarters_address_state", "office_address_state"],
```
`preparer_email` should also get excluded here, emails are case insensitive

seeess1 replied:
Updated

e-belfer reviewed (changes requested):
A few minor comments, but I think you're basically ready to move on to writing this to the PUDL DB. The next steps should be outlined in the issue, but obviously let me know if you have questions about anything!

e-belfer commented on src/pudl/workspace/datastore.py:
```diff
@@ -206,7 +206,7 @@ class ZenodoDoiSettings(BaseSettings):
     ferc714: ZenodoDoi = "10.5281/zenodo.13149091"
     gridpathratoolkit: ZenodoDoi = "10.5281/zenodo.10892394"
-    phmsagas: ZenodoDoi = "10.5281/zenodo.10493790"
+    phmsagas: ZenodoDoi = "10.5281/zenodo.13624526"
```
@seeess1 One more thing to flag - we published new PHMSA data this month, so the DOI should now be:

```suggestion
    phmsagas: ZenodoDoi = "10.5281/zenodo.14026420"
```

seeess1 replied:
Updated

seeess1 commented on src/pudl/helpers.py:
```diff
@@ -2189,111 +2203,105 @@ def retry(
-    and does not perform any data transformation or cleaning.
-    This function checks each column for missing or custom missing values and prints
-    a summary of the findings for string (object), numeric, and datetime columns.
+def analyze_missing_values(
```
Yep all of that is correct. This was just a tool for analysis that could be used in the middle of dev work. 

But I honestly don't feel too precious about it. I'll add a note in the docstring now to make sure that it's clear this is just an analytical tool and not one used in any data transformation. But I'm also totally okay if you want to drop it entirely.

seeess1 replied:
Updated

seeess1 commented on src/pudl/metadata/enums.py:
```diff
@@ -640,3 +640,11 @@
+        & (POLITICAL_SUBDIVISIONS.subdivision_type == "state"),
+        "subdivision_name",
+    ]
+)
```
Updated

seeess1 commented on src/pudl/metadata/fields.py:
```diff
@@ -4249,6 +4250,7 @@
         # different tables. E.g. state of the utility's HQ address vs. state that a
         # plant is located in vs. state in which a utility provides service.
         "description": "Two letter US state abbreviation.",
+        "constraints": {"enum": US_STATE_CODES},
```
I'm going to work on this part next.

seeess1 commented on src/pudl/helpers.py:
```diff
@@ -1009,19 +1009,33 @@ def remove_leading_zeros_from_numeric_strings(
+    df = df.replace(na_patterns, np.nan, regex=True)
+
+    # Attempt to infer original types where possible to avoid unwanted type changes
+    df = df.infer_objects()
```
Updated

e-belfer commented on src/pudl/helpers.py:
```diff
@@ -1009,19 +1009,33 @@ def remove_leading_zeros_from_numeric_strings(
     Currently replaces empty string, single decimal points with no numbers,
-    any single whitespace character, and hyphens with np.nan.
+    any single whitespace character, and hyphens with np.nan. Also converts
+    columns to their original types where possible (using 'infer_objects')
```
Ah this was vaguely in reference to my comment about `convert_dtypes()` vs `infer_objects()`, sorry that wasn't clearer!

seeess1 replied:
Whoops! Looks like my documentation was misleading. No call is made to `infer_objects` in this method. I've removed that part of the function description.

zaneselvans commented:
I could see the change of adding an `ENUM` constraint on the `state` field being complicated since it changes the type to `categorical` rather than `string` which may make it hard to work with different `state` columns in different tables together. If that's a frustration or a hold up then I think just cleaning the column and then leaving it unconstrained (until we can come back and really deal with `state` across the board) is also totally fine.

seeess1 commented:
Hey @e-belfer! Made some more updates (commit [here](https://github.com/catalyst-cooperative/pudl/pull/3929/commits/6f504e5a2247bf2f6a12043a4d6ede09f7e891f6)):

- Updated `.init` files.
- Added a new PHMSA resource in `pudl.metadata.resources`.
- Added new fields.

I think the last couple things are for me to:

- Define the PK (more on that below).
- Create a migration.
- Run all tests to make sure nothing is off.

Does that sound right?

**Primary keys**

Regarding primary keys, I figured that the logical PK identifier for this asset would be the combined `operator_id_phmsa` and `report_number`. But apparently there are a few rows with the same values here but slightly different metadata (for example, different office addresses for each row). I've attached the output from filtering down to rows that have the same combined values for these columns. Let me know how you'd like to handle these.
[duplicates.csv](https://github.com/user-attachments/files/18049110/duplicates.csv)

e-belfer commented:
@seeess1 Always some annoying duplicates! Low stakes here since all the actual data is luckily identical, but thanks for catching this. Some thoughts:

- The last four look completely identical to me and could get handled by `df.drop_duplicates()` unless I'm missing some tiny difference?
- For Williamsville, the one that doesn't have any "testing" in the names will be the one we want.
- For the other 2 cases, would prefer to just have a consistent rule and in this case the first row for each looks intuitively more correct (both city addresses actually match the system they're located in). Would be good to add a tiny assertion in here that checks that you aren't dropping more than the expected number of rows of data (2) in case this becomes a huge problem in the next year of data, e.g.

seeess1 commented on src/pudl/helpers.py:
```diff
@@ -2264,14 +2259,17 @@ def analyze_missing_values(
     Args:
         df: The DataFrame to analyze.
-        custom_missing_values: Optional list of custom values to consider
+    custom_missing_values: Optional list of custom values to consider
```
Updated

seeess1 commented:
> Looking on track! I can do a final review once the last little bits are in place, but your remaining to-do list looks right to me.

Pushed up my latest updates. Let me know what you think.

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -219,10 +318,31 @@ def _check(df):
+        return AssetCheckResult(passed=True)
 
-        if error:
+    @asset_check(asset=spec.asset, blocking=True)
```
Not entirely following the logic here - the asset check will run on the `df` after you've dropped these rows in the transform step to create a unique PK, so the `non_unique_groups` returned in this asset check should always be 0. Rather than writing an asset check here, we'll want to actually embed this assertion in the code itself. 

Could be as simple as something like this in the actual filtering functions you've written:
```
assert unfiltered_df.shape[0] - filtered_df.shape[0]  <=3, "Dropped more than expected number of rows due to non-unique groups. Rows dropped: {dropped_rows}
```

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -179,16 +179,113 @@ def core_phmsagas__yearly_distribution_operators(
+def filter_if_test_in_address(group: pd.DataFrame) -> pd.DataFrame:
+    """Filters out rows with "test" in address columns. The logic is as follows:
+
+    1. For any group of rows with the same combination of "operator_id_phmsa"
+       and "report_number":
+        - If at least one row in the group does not contain the string "test"
+          (case-insensitive) in either "office_address_street" or
+          "headquarters_address_street", keep only the rows in the group
+          that do not contain "test" in these columns.
+        - If all rows in the group contain "test" in either of the columns,
+          leave the group unchanged.
```
The indentation here is causing `make docs-build` to fail. Should look like:

```suggestion
    1. For any group of rows with the same combination of "operator_id_phmsa"
       and "report_number":
       - If at least one row in the group does not contain the string "test"
       (case-insensitive) in either "office_address_street" or
       "headquarters_address_street", keep only the rows in the group
       that do not contain "test" in these columns.
       - If all rows in the group contain "test" in either of the columns,
       leave the group unchanged.
```

e-belfer commented on src/pudl/transform/phmsagas.py:
```diff
@@ -219,10 +318,31 @@ def _check(df):
         return AssetCheckResult(passed=True)
 
-    return _check
+    return [_check_percent_unaccounted_for_gas, _check_pk_deduplication]
```
Right now these asset checks aren't loaded into dagster, at least partially because the function to produce them isn't called anywhere but possibly also because you're returning a list a checks here rather than just one check. If we aren't parameterizing the checks (e.g., specifying expected values by year) then looking at the way in which multiple checks are defined in `eia860.py` would be a better analogy than the FERC 714 model.

e-belfer reviewed (changes requested):
Right now the pandera schema check is failing when I run this locally, and I'm not totally sure why based on the metadata returned from the check failure.

Are you seeing the same issue on your end? I'd suggest getting this into alembic and written to the DB and then we can try to debug the problem, as we have some dtype enforcement built into the IO managers.

Otherwise, one recommendation about the other asset checks - which is to simplify them, as I think the way you've set this up won't work conceptually, seeing as the check gets called after the problematic data has already been dropped. Let me know if that all makes sense?

(As always, just let me know if you need any additional support on my end to get this over the line!)

e-belfer commented:
@seeess1 Thanks for all your work on this one! I'm closing this and picking up your work in #4005 to get this over the last few hurdles!
